### ISSUE TITLE
Better way to force killing all pending coroutines when main function finished

### ISSUE THREAD

**Issue by zonyitoo** [zonyitoo/coio-rs#35]
Move from #32 .

**Comment by zonyitoo** [comment]
One possible solution is when the current thread is panicking, then push the current coroutine into a queue, which will be processed after all the worker threads are finished. Pseudocode:

```rust
static GLOBAL_QUEUE: Queue = ...;

struct Handle(*mut Coroutine);

impl Drop for Handle {
    fn drop(&mut self) {
        if thread::panicking {
            GLOBAL_QUEUE.push(Handle(self.0));
        } else {
            let coro = unsafe { Box::from_raw(self.0) };
            drop(coro);
        }
    }
}

impl Scheduler {
    pub fn run(self) -> Result {
        // ...

        let result = main_handle.join();

        // After it gets back

        // 1. Drop all the coroutines that are waiting in EventLoop
        // 2. Stop all the worker threads
        // 3. Drop until the queue is empty
        while let Some(coro) = GLOBAL_QUEUE.pop() {
            drop(coro);
        }

        result
    }
}
```

**Comment by zonyitoo** [comment]
This proposal also helped to find the bug in #36 . And it is working now. LoL.

**Comment by zonyitoo** [comment]
The old implementation is dropped because we found that `panic::propagate` won't trigger "panic while panicking". So right now we don't need to maintain the drop queue.

**Comment by zonyitoo** [comment]
Merged in e1d124718f451a6f04d1babb35c0452c1b32e933
